storage/transfermanager/downloader.go

```diff
@@ -156,46 +175,112 @@ func (d *Downloader) downloadWorker() {
 			break // no more work; exit
 		}
 
-		// TODO: break down the input into smaller pieces if necessary; maybe as follows:
-		// Only request partSize data to begin with. If no error and we haven't finished
-		// reading the object, enqueue the remaining pieces of work and mark in the
-		// out var the amount of shards to wait for.
-		out := input.downloadShard(d.client, d.config.perOperationTimeout)
-
-		// Keep track of any error that occurred.
-		if out.Err != nil {
-			d.error(fmt.Errorf("downloading %q from bucket %q: %w", input.Object, input.Bucket, out.Err))
-		}
-
-		// Either execute the callback, or append to results.
-		if d.config.asynchronous {
-			input.Callback(out)
+		out := input.downloadShard(d.client, d.config.perOperationTimeout, d.config.partSize)
+
+		if input.shard == 0 {
+			if out.Err != nil {
+				// Don't queue more shards if the first failed.
+				d.addResult(input, out)
+			} else {
+				numShards := numShards(out.Attrs, input.Range, d.config.partSize)
+
+				if numShards <= 1 {
+					// Download completed with a single shard.
+					d.addResult(input, out)
+				} else {
+					// Queue more shards.
+					outs := d.queueShards(input, out.Attrs.Generation, numShards)
+					// Start a goroutine that gathers shards sent to the output
+					// channel and adds the result once it has received all shards.
+					go d.gatherShards(input, outs, numShards)
+				}
+			}
 		} else {
-			d.addResult(out)
+			// If this isn't the first shard, send to the output channel specific to the object.
+			// This should never block since the channel is buffered to exactly the number of shards.
+			input.shardOutputs <- out
 		}
 	}
 	d.workers.Done()
 }
 
+// queueShards queues all subsequent shards of an object after the first.
+// The results should be forwarded to the returned channel.
+func (d *Downloader) queueShards(in *DownloadObjectInput, gen int64, shards int) <-chan *DownloadOutput {
+	// Create a channel that can be received from to compile the
+	// shard outputs.
+	outs := make(chan *DownloadOutput, shards)
+	in.shardOutputs = outs
+
+	// Create a shared context that we can cancel if a shard fails.
+	in.ctx, in.cancelCtx = context.WithCancelCause(in.ctx)
+
+	// Add generation in case the object changes between calls.
+	in.Generation = &gen
+
+	// Queue remaining shards.
+	// We could optimize for zero-initialized DownloadBuffers if we queue the last
```

Comment:
Maybe turn this into a TODO for profiling work?

Reply:
Moved it into perf ticket